### PR TITLE
Add option to aggressively inline all literals and consts

### PR DESCRIPTION
**Pull request by jwatzman**
This new inlining option does two things:
- Inlines any variable which looks like a literal constant
  (`hasInitDeps` is false) and which does not appear in an lvalue
  position
- Inlines any variable marked as `const` whatsoever.

This can increase program size, if a large literal is inlined over and
over. However, the idea is exactly why the Google Closure Compiler also
always inlines constants: a (the?) major use-case of this minifier is
for demoscene shaders, which are almost certainly going to be put
through compression later anyway -- so what matters is the size after
compression, not the actual output size. And reducing the redudancy that
this inlining creates can be done by the compressor in way fewer bytes
than declaring a new variable in the program text.

https://github.com/google/closure-compiler/wiki/FAQ#closure-compiler-inlined-all-my-strings-which-made-my-code-size-bigger-why-did-it-do-that

This option isn't the default; we likely want to do more testing to see
if it actually works out that way in practise (and to make sure the
output doesn't break things!) However, when I tested it on the two
largest shaders in `tests/real` (`the_real_party...` and `lunaquatic`),
it not only reduced gzip'd size but also raw output size too. So it may
just be a win all-around anyway?

- `the_real_party...`: 17708 -> 17699 (gzipped 5181 -> 5173)
- `lunaquatic`: 7356 -> 7347 (gzipped 2745 -> 2739)

### CODE REVIEW

File: src/rewriter.fs [hunk 1]
```diff
@@ -285,8 +291,79 @@ let rec iterateSimplifyAndInline li =
...
+    let findWrites env = function
+        | FunCall(Op o, e::args) when Set.contains o assignOps ->
+            let newEnv = {env with fExpr = markVars}
+            FunCall(Op o, (mapExpr newEnv e)::args)
+        | FunCall(Var v, _) as e ->
+            let fnIsMarked =
+                match env.fns.TryFind v.Name with
+                | Some fct when fct.fName.IsLValue -> true
+                | _ -> false
+            if fnIsMarked then
```
Comment (laurentlb) [inline src/rewriter.fs:337]: I don't think you need fnIsMarked:

```f#
match env.fns.TryFind v.Name with
    | Some fct when fct.fName.IsLValue ->
        let newEnv = {env with fExpr = markVars}
        mapExpr newEnv e
    | _ -> e
```

**Review by laurentlb — COMMENTED** [review]
Thanks a lot!

File: src/rewriter.fs [hunk 2]
```diff
@@ -285,8 +291,79 @@ let rec iterateSimplifyAndInline li =
     let simplified = mapTopLevel (mapEnv simplifyExpr simplifyStmt) li
     if foundInlinable then iterateSimplifyAndInline simplified else simplified
 
+let inlineAllConsts li =
+    let mapInnerDecl = function
+        // Unconditional inlining of anything marked "const" -- trust that the
+        // compiler would have yelled if it weren't really really const, so we
+        // can brutishly just inline it.
+        | ({typeQ = tyQ}, defs) as d when List.contains "const" tyQ ->
+            List.iter (fun (def:DeclElt) -> def.name.Inline()) defs
```
[the remark this answers is not in the document]
Comment (jwatzman) [reply src/rewriter.fs:300]: Ah cool, yeah that reads better to me too. I don't think OCaml has this construction, and that's the language that I'm far and away the more familiar with, so good to learn.

File: src/rewriter.fs (hunk 1 above)
Comment (jwatzman) [reply src/rewriter.fs:337]: Er, yes, of course, not sure what I was thinking here.

**Review by laurentlb — APPROVED** [review]

**Pull request merged**